src/psyclone/psyir/nodes/routine.py

```diff
@@ -64,14 +71,21 @@ class Routine(Schedule, CommentableMixin):
     _children_valid_format = "[Statement]*"
     _text_name = "Routine"
 
-    def __init__(self, name, is_program=False, **kwargs):
+    def __init__(self, symbol, is_program=False,
+                 symbol_tag=None, **kwargs):
+        # These attributes need to be set before anything, as the _symbol
+        # is required for setting the _parent, which is overriden by Routine
+        if not isinstance(symbol, RoutineSymbol):
+            raise TypeError(f"Routine argument 'symbol' must be present and "
+                            f"must be a RoutineSymbol but got "
+                            f"'{type(symbol).__name__}'")
+        self._symbol = symbol
+        self._symbol_tag = symbol_tag
+        self._symbol_in_table = False
+        self._parent_node = None
```

Comment:
Yes - since `_parent` is a function and not a variable the actual `_parent` is stored in `_parent_node`.

Reply:
Comment added

Reply:
I believe you can still access the parent variable with super()._parent, but maybe it will be clearer if instead of overloading the properties we just did:
```
  def __setattr__(self, name, value):
      if name == '_parent':
          self.update_parent_symbol_table(value)
      super().__setattr__(name, value)

  def update_parent_symbol_tables(self, new_parent):
      """ Remove self._symbol from self._parent.symbol_table and add it to new_parent.symbol_table """
```

Alternatively, we could just added this method to the places where "_parent" is meant to be set which are all at Node  (in `set_parent_link`, `del_parent_link`, `__init__`, and `_refine_copy`) e.g.:
```
        node.update_parent_symbol_tables(self._node_reference)
        node._parent = self._node_reference
``` 
It is not as automatic but, we would still have `routine._parent =` if we would want to do something nasty in a particular situation or test and really force the reference change without the rest of the logic.

What do you think?

Reply:
You can find the parent with `._parent` but I think whenever you do `._parent = X` it would call the setter (so the property and setter reference ._parent_node as the actual storage location) - at least this is my understanding of how it can work.

I will try the first implementation you suggested and unless anything breaks we can use that :)

I don't think we can just override those functions in node as `ChildrenList` does `._parent` setting directly.

Reply:
This change did break some things in copying and initialization, so I had to add a little code to `update_p;arent_symbol_table` to handle that too.